src/App.tsx

```diff
@@ -1,9 +1,241 @@
+  line-height: 0.875rem;
+`;
+
+const Wrapper = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 8px;
+`;
 
 function App() {
+  const { cardNumbers, cardNumbersChangeHandler } = useCardNumber(
+    Array.from({ length: 4 }, () => InitialCardNumberState),
+  );
+  const [cardImageSrc, setCardImageSrc] = useState('');
+
+  const {
+    inputState: month,
+    inputChangeHandler: monthChangeHandler,
+    error: monthError,
+    setError: setMonthError,
+  } = useInput();
+
+  const {
+    inputState: year,
+    inputChangeHandler: yearChangeHandler,
+    error: yearError,
+    setError: setYearError,
+  } = useInput();
+
+  const {
+    inputState: name,
+    inputChangeHandler: nameChangeHandler,
+    error: nameError,
+    setError: setNameError,
+  } = useInput();
+
+  useEffect(() => {
+    if (
+      month !== '' &&
+      (!validate.isNumberInRange({ min: 1, max: 12, compareNumber: Number(month) }) ||
+        !validate.isValidDigit(month))
+    ) {
+      setMonthError(true);
+
+      return;
+    }
+
+    setMonthError(false);
+  }, [month, monthError]);
+
+  useEffect(() => {
+    if (year !== '' && !validate.isValidDigit(year)) {
+      setYearError(true);
+
+      return;
+    }
+
+    setYearError(false);
+  }, [year, yearError]);
+
+  useEffect(() => {
+    if (name !== '' && !validate.isEnglish(name)) {
+      setNameError(true);
+
+      return;
+    }
+
+    setNameError(false);
+  }, [name, nameError]);
+
+  useEffect(() => {
+    setCardImageSrc('');
+
+    const cardNumberString = cardNumbers.map((cardNumber) => cardNumber.value).join('');
+
+    if (validate.isVisa(cardNumberString)) {
+      setCardImageSrc(VisaCardImage);
+    }
+
+    if (validate.isMasterCard(cardNumberString)) {
+      setCardImageSrc(MasterCardImage);
+    }
+  }, [cardNumbers]);
+
   return (
     <>
-      <h1>React Payments</h1>
+      <GlobalStyles />
+      <Container>
+        <CreditCard
+          cardNumbers={cardNumbers}
+          month={month}
+          year={year}
+          name={name}
+          cardImageSrc={cardImageSrc}
+        />
+        <CardInfoContainer>
+          <Wrapper>
+            <InputInfo
+              title={CARD_NUMBER.title}
+              description={CARD_NUMBER.description}
+              inputTitle={CARD_NUMBER.inputTitle}
+            >
+              {cardNumbers.map((cardNumber, index) => {
+                const uniqueId = 'cardNumbers' + index;
+                return (
+                  <>
+                    <Label htmlFor={uniqueId} />
+                    <Input
+                      key={uniqueId}
+                      id={uniqueId}
+                      placeholder="1234"
+                      type="text"
+                      maxLength={4}
+                      value={cardNumber.value}
+                      onChange={(e) => cardNumbersChangeHandler(e, index)}
+                      isError={cardNumber.isError}
+                    />
+                  </>
```

Comment:
`key`는 맵 대상에 지정해야할 것 같습니다. `key={'cardNumbers' + index}` 와 `key={index}`에 어떤 차이가 있을까요?

```suggestion
                  <React.Fragment key={index}>
                    <Label htmlFor={uniqueId} />
                    <Input
                      id={uniqueId}
                      placeholder="1234"
                      type="text"
                      maxLength={4}
                      value={cardNumber.value}
                      onChange={(e) => cardNumbersChangeHandler(e, index)}
                      isError={cardNumber.isError}
                    />
                  </>
```

Reply:
아하 맞는 말씀입니다..!

짚어주셔서 감사합니다 🙇 

저 `key`값이 리액트 render phase 과정에서 업데이트를 시켜야할지 말지를 판단하는 값중에 하나라고 알고 있는데요..!

그래서 만약 단순히 index값으로 해주는 경우에, map함수로 뿌려준 요소들 중에서 삭제나 추가 같은 이벤트가 발생하면 index 들이 전부 수정이 되어버리기때문에 전부 리렌더링이 일어나는거로 알고 있습니다! 

혹시 이러한 부분에 대해서 질문 주신것이 맞으실까요 ?

혹시나 잘못 생각하고 있거나, 추가로 말씀 주실 부분이 있다면 짚어주시면 감사하겠습니다!

Reply:
버건디가 이해하신 의도로 질문드렸던 것이 맞습니다 :) 

제가 덧붙여 설명 드려보자면, index를 `key`로 그대로 사용하는 것이 언제나 틀린 것이 아닙니다. 

key의 사용목적은 **key'는 해당 항목의 고유성을 보장해줘서, 리액트가 항목의 추가, 삭제 또는 변경하기 위해 해당 항목을 식별**하기 위함입니다.

따라서, 리스트에 변경사항이 없다면 index를 그대로 사용해도 무방하지요.

또한 key는 sibling끼리만 고유하면 되므로, 여기 코드에서 key={'cardNumbers' + index} 와 key={index}에는 실질적인 차이가 없어 보입니다. (그래서 질문을 드리게 되었어요.)

> 단순히 index값으로 해주는 경우에, 삭제나 추가 같은 이벤트가 발생하면 index 들이 전부 수정이 되어버리기때문에 전부 리렌더링이 일어나는거로 알고 있습니다!

리렌더링이 전부 (의도한 대로) 일어나면 다행인데, 그렇지 않은 문제가 발생할 수 있어 주의해야합니다. 더 자세한 내용은 공식문서를 참고해주시면 좋겠습니다 👍

- 참고: https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key

Reply:
>또한 key는 sibling끼리만 고유하면 되므로, 여기 코드에서 key={'cardNumbers' + index} 와 key={index}에는 실질적인 차이가 없어 보입니다. (그래서 질문을 드리게 되었어요.)

헉 맞네요..🥲 step2때 반영하도록 하겠습니다! 짚어주셔서 감사해요 하루~!